### DISCUSSION TITLE
link module with external library

### DISCUSSION THREAD

**Discussion by sheershak-agarwal** [nasa/fprime#1078] [Q&A] [answered]
Hello,

I am trying to add armadillo library to a component in my f-prime design.

I can compile my test application with "g++ arma.cpp -larmadillo" command.
I can see the library in "/usr/lib/" as libarmadillo.so   

How do I include it in my component? I tried adding this to my components cmakelists.txt file
```
get_module_name("${CMAKE_CURRENT_LIST_DIR}")
include_directories("/usr/lib/")
find_library(ARMADILLO armadillo "/usr/lib/armadillo.so")
target_link_libraries(${MODULE_NAME} ${ARMADILLO})
```
but it didn't work.

**Comment by astroesteban** [comment]
**What errors are you seeing?**

Also this discussion might be of help to you:
https://github.com/nasa/fprime/discussions/976

**Reply by sheershak-agarwal** [reply]
I did see the thread and copied the above lines from its solution. My library is C++ so I don't need to worry about extern C
I get this error:
```
The plain signature for target_link_libraries has already been used with
See also "/home/sheershak/Desktop/F-Prime-Debug/App/build-fprime-automatic-raspberrypi/CMakeFiles/CMakeError.log".
  the target "App_AdcsManager".  All uses of target_link_libraries with a
Makefile:2287: recipe for target 'cmake_check_build_system' failed
  target must be either all-keyword or all-plain.

  The uses of the plain signature are here:

   * /home/sheershak/Desktop/F-Prime-Debug/fprime/cmake/support/Module.cmake:50 (target_link_libraries)
   * /home/sheershak/Desktop/F-Prime-Debug/fprime/cmake/support/Module.cmake:95 (target_link_libraries)



make: *** [cmake_check_build_system] Error 1
[ERROR] CMake erred with return code 2
```

when I am using this inside my component's Cmakelists:
```
set(SOURCE_FILES
	"${CMAKE_CURRENT_LIST_DIR}/AdcsManagerComponentAi.xml"
	"${CMAKE_CURRENT_LIST_DIR}/AdcsManagerComponentImpl.cpp"
)
set(MOD_DEPS
  "App/Common"
)
register_fprime_module()
get_module_name("${CMAKE_CURRENT_LIST_DIR}")

include_directories("/usr/lib/")
find_library(ARMADILLO armadillo "/usr/lib/armadillo.so")

# target_link_libraries(${MODULE_NAME} ${ARMADILLO})
target_link_libraries(${MODULE_NAME} PUBLIC ARMADILLO)
```

**Reply by sheershak-agarwal** [reply]
I can see my libarary in /usr/lib/ as:
```
libarmadillo.so
libarmadillo.so.8
libarmadillo.so.8.400.0
```

**Reply by sheershak-agarwal** [reply]
I also tried this code:
```
get_module_name("${CMAKE_CURRENT_LIST_DIR}")

# include_directories("/usr/lib/")
include_directories("${PROJECT_SOURCE_DIR}/lib")
find_library(LIBARMADILLO armadillo "/usr/lib/")

# target_link_libraries(${MODULE_NAME} ${ARMADILLO})
target_link_libraries("${MODULE_NAME}" ${LIBARMADILLO})
```
but found this error:
```
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
-- Generating done
Please set them or make sure they are set and tested correctly in the CMake files:
LIBARMADILLO
    linked by target "App_AdcsManager" in directory /home/sheershak/Desktop/F-Prime-Debug/App/AdcsManager

CMake Generate step failed.  Build files cannot be regenerated correctly.
Makefile:2287: recipe for target 'cmake_check_build_system' failed
make: *** [cmake_check_build_system] Error 1
[ERROR] CMake erred with return code 2
```

**Comment by sheershak-agarwal** [comment, marked as answer]
Hello, I am still stuck on this part.

**Reply by ThibFrgsGmz** [reply]
Have you tried @astroesteban's "next-gen" method in the thread he mentions?

Since everyone seems to be busy, and despite I'm far from being a CMake expert, I'll do my best to support you.

From what I've seen on StackOverflow, the procedure could be as follows.


```cmake
# Find all the libraries contained in the directory you specified.
find_library(LIBARMADILLO armadillo "/usr/lib")

# GLOBAL should be added if the imported library is located in directories above the current.
add_library(armadillo SHARED IMPORTED GLOBAL)
set_property(TARGET armadillo PROPERTY IMPORTED_LOCATION ${LIBARMADILLO})
```
Or,
```cmake
# Set libraries search path:
link_directories(/usr/lib)

# Link library:
target_link_libraries(LIBARMADILLO armadillo)
```


I don't have time to test it right now, but maybe by the end of the week.

**Reply by astroesteban** [reply]
Sorry I've been pretty busy with work dealing with the same issue of linking against pre-compiled libraries in CMake. As @ThibFrgsGmz mentioned try that approach:
```cmake
# Define the armadillo CMake target
add_library(armadillo SHARED IMPORTED GLOBAL)
set_target_properties(armadillo PROPERTIES IMPORTED_LOCATION /usr/lib/armadillo.so)  # <-- You have to give the library name too
target_include_directories(armadillo SYSTEM INTERFACE /path/to/any/headers/armadillo/needs)  # <-- optional
```
The most important part is the `GLOBAL` keyword. Without it imported targets are limited to the current CMakeLists that's being processes. That means anything that uses `armadillo` outside of where you declare it can't see it. The `GLOBAL` exposes the library to the entire CMake build.

**Reply by sheershak-agarwal** [reply]
Thank you @ThibFrgsGmz and @astroesteban. the code below works

```
find_library(LIBARMADILLO armadillo "/usr/lib")

# GLOBAL should be added if the imported library is located in directories above the current.
add_library(armadillo SHARED IMPORTED GLOBAL)
set_property(TARGET armadillo PROPERTY IMPORTED_LOCATION ${LIBARMADILLO})
```

**Reply by teddylambert** [reply]
I'm having a similar issue with the `gpiod` library. Running `dpkg -L libgpiod-dev` returns:
```
/.
/usr
/usr/include
/usr/include/gpiod.h
/usr/include/gpiod.hpp
/usr/lib
/usr/lib/x86_64-linux-gnu
/usr/lib/x86_64-linux-gnu/libgpiod.a
/usr/lib/x86_64-linux-gnu/libgpiodcxx.a
/usr/lib/x86_64-linux-gnu/pkgconfig
/usr/lib/x86_64-linux-gnu/pkgconfig/libgpiod.pc
/usr/lib/x86_64-linux-gnu/pkgconfig/libgpiodcxx.pc
/usr/share
/usr/share/doc
/usr/share/doc/libgpiod-dev
/usr/share/doc/libgpiod-dev/README.gz
/usr/share/doc/libgpiod-dev/copyright
/usr/lib/x86_64-linux-gnu/libgpiod.so
/usr/lib/x86_64-linux-gnu/libgpiodcxx.so
/usr/share/doc/libgpiod-dev/changelog.Debian.gz
```

When I try to use the above fix
```
find_library(LIBGPIOD libgpiod "/usr/lib/x86_64-linux-gnu")
add_library(libgpiod SHARED IMPORTED GLOBAL)
set_target_properties(libgpiod PROPERTIES IMPORTED_LOCATION /usr/lib/x86_64-linux-gnu/libgpiod.so)
```
I still end up with a bunch of `undefined reference to <gpiod library function>` around 100% build step which I think means it's still not linking correctly?

Is there something else I'm missing or doing incorrectly? The library is used by a specific component of my deployment, so I have the CMake lines in the component `CMakeLists.txt` below the `register_fprime_module()` line. The hpp file of the component includes `gpiod.h`, and at least that part seems to work correctly.

**Reply by LeStarch** [reply]
You need to ensure that you include a link to the library itself.  

```
set(MOD_DEPS
    libgpiod
    ...
)
register_fprime_module()
```

**Reply by teddylambert** [reply]
Got it, had missed that.

I now have it set up as:
```
set(MOD_DEPS
    libgpiod
)

register_fprime_module()

find_library(LIBGPIOD libgpiod "/usr/lib/x86_64-linux-gnu")
add_library(libgpiod SHARED IMPORTED GLOBAL)
set_target_properties(libgpiod PROPERTIES IMPORTED_LOCATION /usr/lib/x86_64-linux-gnu/libgpiod.so)
```
and am met with a `/usr/lib/x86_64-linux-gnu/libgpiod.so: error adding symbols: File in wrong format` error. I checked the library header file (`gpiod.h`) and it has
```
#ifdef __cplusplus
extern "C" {
#endif
```
Is there something else I need to include?

**Reply by LeStarch** [reply]
File in wrong format implies the library you are linking against is not the same as the one you are building.

1. Are you cross-compiling?  If so, is your target `x86_64`?  
2. Are you building locally on a machine which is not `x86_64`?

For example, if you are building for the Raspberry PI (arm) and you are linking against the above library (x86_64) it will not link.  You'll need to find an arm build of the library.
